### PR TITLE
Add a tombstone operation to optimize multiple deletes

### LINKED ISSUE #31699: Optimize Remote Infinispan performance on removal of entry
### Description

The current implementation optimizes that a put and a removal cancel out each other so they will never touch the remote cache:

https://github.com/keycloak/keycloak/blob/9006218559496d5fb6fe98cd9760cad7f57b62e7/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/remote/RemoteInfinispanKeycloakTransaction.java#L102-L106

This doesn't handle cases where the put it called once, and the remove is called multiple times. This is identical to the optimizations done in #28672

### Discussion

_No response_

### Motivation

Improve the performance of the remote cache as it has shown a regression in response time performance in https://github.com/keycloak/keycloak-benchmark/issues/901#issuecomment-2252501370 for cpuUsageForCredentialGrantsTest

### Details

I'll create a PR

### PR DESCRIPTION
**Pull request by ahus1**
Closes #31699

### CODE REVIEW

**Comment by ahus1** [comment]
@pruivo - this PR will optimize the performance on client credential grants. It will also slightly optimize the logout as there also a temporary authentication session is created. 

Please give it a review. Thanks!

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/remote/RemoteInfinispanKeycloakTransaction.java [hunk 1]
```diff
@@ -254,4 +256,18 @@ public CompletionStage<?> execute(RemoteCache<K, V> cache) {
             return cache.removeAsync(key);
         }
     }
+
+    private record TombstoneOperation<K, V>() implements Operation<K, V> {
```
Comment (pruivo) [inline model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/remote/RemoteInfinispanKeycloakTransaction.java:260]: can be a constant
```java
    private static final Operation<?,?> TOMBSTONE = new Operation<>() {

        @Override
        public boolean canRemove() {
            return true;
        }

        @Override
        public CompletionStage<?> execute(RemoteCache<Object, Object> cache) {
            return CompletableFutures.completedNull();
        }
    };
```

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/remote/RemoteInfinispanKeycloakTransaction.java [hunk 2]
```diff
@@ -254,4 +256,18 @@ public CompletionStage<?> execute(RemoteCache<K, V> cache) {
...
+
+        @Override
+        public boolean canRemove() {
+            return true;
+        }
+
+        @Override
+        public CompletionStage<?> execute(RemoteCache<K, V> cache) {
+            return CompletableFuture.completedStage(null);
```
Comment (pruivo) [inline model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/remote/RemoteInfinispanKeycloakTransaction.java:269]:
```suggestion
            return CompletableFutures.completedNull();
```

**Review by pruivo — CHANGES_REQUESTED** [review]

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/remote/RemoteInfinispanKeycloakTransaction.java (hunk 1 above)
Comment (ahus1) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/remote/RemoteInfinispanKeycloakTransaction.java:261]: Thanks, updated in next commit. I used `org.infinispan.commons.util.concurrent.CompletableFutures` for this.

**Review by pruivo — APPROVED** [review]
LGTM.

**Review by mhajas — APPROVED** [review]
LGTM @ahus1

**Comment by ahus1** [comment]
@pruivo - the logic stayed the same, I only needed to change the IF condition slightly to make all tests pass (there was a situation where there was a remove queue and then a put wanted to be added. This is now consolidated as just a put).

**Pull request merged**